File: ios/Sources/MeasureSDK/Swift/Tracing/MsrTracer.swift
//
//  MsrTracer.swift
//  Measure
//
//  Created by Adwin Ross on 10/04/25.
//

import Foundation

final class MsrTracer: Tracer {
    private let logger: Logger
    private let idProvider: IdProvider
    private let timeProvider: TimeProvider
    private let spanProcessor: SpanProcessor
    private let sessionManager: SessionManager
    private let signalSampler: SignalSampler

    init(logger: Logger,
         idProvider: IdProvider,
         timeProvider: TimeProvider,
         spanProcessor: SpanProcessor,
         sessionManager: SessionManager,
         signalSampler: SignalSampler) {
        self.logger = logger
        self.idProvider = idProvider
        self.timeProvider = timeProvider
        self.spanProcessor = spanProcessor
        self.sessionManager = sessionManager
        self.signalSampler = signalSampler
    }

    func spanBuilder(name: String) -> SpanBuilder {
        return MsrSpanBuilder(name: name,
                              idProvider: idProvider,
                              timeProvider: timeProvider,
                              spanProcessor: spanProcessor,
                              sessionManager: sessionManager,
                              signalSampler: signalSampler,
                              logger: logger)
    }

    func getTraceParentHeaderValue(for span: Span) -> String {
        let sampledFlag = span.isSampled ? "01" : "00"
        return "00-\(span.traceId)-\(span.spanId)-\(sampledFlag)"
    }

    func getTraceParentHeaderKey() -> String {
        return "traceparent"
    }
}

File: android/measure-android/measure/src/main/java/sh/measure/android/tracing/MsrTracer.kt
package sh.measure.android.tracing

import sh.measure.android.SessionManager
import sh.measure.android.logger.Logger
import sh.measure.android.utils.IdProvider
import sh.measure.android.utils.Sampler
import sh.measure.android.utils.TimeProvider

internal class MsrTracer(
    private val logger: Logger,
    private val idProvider: IdProvider,
    private val timeProvider: TimeProvider,
    private val spanProcessor: SpanProcessor,
    private val sessionManager: SessionManager,
    private val sampler: Sampler,
) : Tracer {
    override fun spanBuilder(name: String): SpanBuilder = MsrSpanBuilder(
        name,
        idProvider,
        timeProvider,
        spanProcessor,
        sessionManager,
        sampler,
        logger,
    )

    override fun getTraceParentHeaderValue(span: Span): String {
        val sampledFlag = if (span.isSampled) "01" else "00"
        return "00-${span.traceId}-${span.spanId}-$sampledFlag"
    }

    override fun getTraceParentHeaderKey(): String = "traceparent"
}
